### PR TITLE
Encryption: fix circular reference in KeyspacesKeyManager

### LINKED ISSUE #8351: Support encryption at rest under disaggregate mode
## Feature Request

**Is your feature request related to a problem? Please describe:**

Encryption at rest provides data protection for stored data (at rest). In disaggregate TiFlash, we still do not support it.

**Describe the feature you'd like:**

Each keyspace, each file uses its own encryption key. 

- Different files use different keys.
- For data belonging to different keyspaces and stored in the same file, should use different keys.

**Describe alternatives you've considered:**

**Teachability, Documentation, Adoption, Migration Strategy:**

### PR DESCRIPTION
**Pull request by Lloyd-Pottiger**
### What problem does this PR solve?

Issue Number: close #8351

Problem Summary:

### What is changed and how it works?

```
file_provider -> key_manager
key_manager -> page_storage
page_storage -> file_provider
```

own `std::weak_ptr<UniversalPageStorage>` in `KeyspacesKeyManager`

### Check List

Tests 

- [ ] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: dbms/src/Encryption/KeyspacesKeyManager.cpp [hunk 1]
```diff
@@ -40,11 +47,11 @@ FileEncryptionInfo KeyspacesKeyManager<ProxyHelper>::getInfo(const EncryptionPat
         };
     }
 
-    RUNTIME_CHECK(ps_write != nullptr);
+    RUNTIME_CHECK(!ps_write.expired());
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/KeyspacesKeyManager.cpp:50]: Should use weak_ptr as follows. Or there is still a chance that the instance is released between you calling `ps_write.expired()` and `ps_write.lock()`.
```
auto ps_write_locked = ps_write.lock();
RUNTIME_CHECK(!ps_write_locked);
...
ps_write_locked->read(...);
```

Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/KeyspacesKeyManager.cpp]: done

**Review by CalvinNeo — APPROVED** [review]
lgtm

File: dbms/src/Encryption/KeyspacesKeyManager.cpp [hunk 2]
```diff
@@ -122,11 +131,12 @@ void KeyspacesKeyManager<ProxyHelper>::deleteKey(KeyspaceID keyspace_id)
     if (unlikely(keyspace_id == NullspaceID) || (likely(!proxy_helper->getKeyspaceEncryption(keyspace_id))))
         return;
 
-    RUNTIME_CHECK(ps_write != nullptr);
+    auto ps_write_locked = ps_write.lock();
+    RUNTIME_CHECK(!ps_write_locked);
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/KeyspacesKeyManager.cpp:135]:
```suggestion
    RUNTIME_CHECK(ps_write_locked != nullptr);
```

File: dbms/src/Encryption/KeyspacesKeyManager.cpp [hunk 3]
```diff
@@ -78,8 +86,9 @@ FileEncryptionInfo KeyspacesKeyManager<ProxyHelper>::newInfo(const EncryptionPat
         };
     }
 
-    RUNTIME_CHECK(ps_write != nullptr);
-    auto load_func = [this, keyspace_id]() -> EncryptionKeyPtr {
+    auto ps_write_locked = ps_write.lock();
+    RUNTIME_CHECK(!ps_write_locked);
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/KeyspacesKeyManager.cpp:90]:
```suggestion
    RUNTIME_CHECK(ps_write_locked != nullptr);
```

File: dbms/src/Encryption/KeyspacesKeyManager.cpp [hunk 4]
```diff
@@ -40,11 +47,12 @@ FileEncryptionInfo KeyspacesKeyManager<ProxyHelper>::getInfo(const EncryptionPat
...
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/KeyspacesKeyManager.cpp:51]:
```suggestion
    RUNTIME_CHECK(ps_write_locked != nullptr);
```

**Review by JaySon-Huang — COMMENTED** [review]
`!ps_write_locked` means `ps_write_locked == nullptr`

File: dbms/src/Encryption/KeyspacesKeyManager.h [hunk 5]
```diff
@@ -62,7 +58,7 @@ class KeyspacesKeyManager : public KeyManager
 public:
 #endif
     ProxyHelper * proxy_helper;
-    UniversalPageStoragePtr ps_write = nullptr;
+    std::weak_ptr<UniversalPageStorage> ps_write;
```
Comment (JaySon-Huang) [inline dbms/src/Encryption/KeyspacesKeyManager.h:61]: Add some comments about why we need a weak_ptr here

**Review by JaySon-Huang — APPROVED** [review]
Rest LGTM

File: dbms/src/Encryption/KeyspacesKeyManager.h (hunk 5 above)
Comment (Lloyd-Pottiger) [reply dbms/src/Encryption/KeyspacesKeyManager.h:65]: done

**Pull request merged**